File: tests/mv-35/discs.pl
#! /usr/bin/perl
use utf8;
use strict;
use warnings;
no warnings "uninitialized";
use re::engine::RE2;
use JSON::JsonModel;
use constant JMC_VERSION => "2";


sub _jm_obj_0($$$);
sub _jm_obj_1($$$);
sub _jm_obj_2($$$);
sub _jm_obj_3($$$);
sub _jm_obj_4($$$);
sub _jm_obj_5($$$);
sub _jm_obj_6($$$);
sub _jm_obj_7($$$);
my %_jm_map_0;
sub json_model_1($$$);
my %check_model_map;

sub _jm_obj_0($$$)
{
    my ($val, $path, $rep) = @_;
    return 0 unless jm_is_object($val);
    return 0 if jm_obj_size($val) != 2;
    my $pval;
    return 0 unless exists $$val{"d"};
    $pval = $$val{"d"};
    my $res = jm_is_string($pval) && $pval eq "n";
    return 0 unless $res;
    return 0 unless exists $$val{"v"};
    $pval = $$val{"v"};
    return !defined($pval);
}

sub _jm_obj_1($$$)
{
    my ($val, $path, $rep) = @_;
    return 0 unless jm_is_object($val);
    return 0 if jm_obj_size($val) != 2;
    my $pval;
    return 0 unless exists $$val{"d"};
    $pval = $$val{"d"};
    my $res = jm_is_string($pval) && $pval eq "b";
    return 0 unless $res;
    return 0 unless exists $$val{"v"};
    $pval = $$val{"v"};
    return jm_is_boolean($pval);
}

sub _jm_obj_2($$$)
{
    my ($val, $path, $rep) = @_;
    return 0 unless jm_is_object($val);
    return 0 if jm_obj_size($val) != 2;
    my $pval;
    return 0 unless exists $$val{"d"};
    $pval = $$val{"d"};
    my $res = jm_is_string($pval) && $pval eq "i";
    return 0 unless $res;
    return 0 unless exists $$val{"v"};
    $pval = $$val{"v"};
    return jm_is_integer($pval);
}

sub _jm_obj_3($$$)
{
    my ($val, $path, $rep) = @_;
    return 0 unless jm_is_object($val);
    return 0 if jm_obj_size($val) != 2;
    my $pval;
    return 0 unless exists $$val{"d"};
    $pval = $$val{"d"};
    my $res = jm_is_string($pval) && $pval eq "u";
    return 0 unless $res;
    return 0 unless exists $$val{"v"};
    $pval = $$val{"v"};
    return jm_is_integer($pval) && $pval >= 0;
}

sub _jm_obj_4($$$)
{
    my ($val, $path, $rep) = @_;
    return 0 unless jm_is_object($val);
    return 0 if jm_obj_size($val) != 2;
    my $pval;
    return 0 unless exists $$val{"d"};
    $pval = $$val{"d"};
    my $res = jm_is_string($pval) && $pval eq "f";
    return 0 unless $res;
    return 0 unless exists $$val{"v"};
    $pval = $$val{"v"};
    return jm_is_numeric($pval);
}

sub _jm_obj_5($$$)
{
    my ($val, $path, $rep) = @_;
    return 0 unless jm_is_object($val);
    return 0 if jm_obj_size($val) != 2;
    my $pval;
    return 0 unless exists $$val{"d"};
    $pval = $$val{"d"};
    my $res = jm_is_string($pval) && $pval eq "s";
    return 0 unless $res;
    return 0 unless exists $$val{"v"};
    $pval = $$val{"v"};
    return jm_is_string($pval);
}

sub _jm_obj_6($$$)
{
    my ($val, $path, $rep) = @_;
    return 0 unless jm_is_object($val);
    return 0 if jm_obj_size($val) != 2;
    my $pval;
    return 0 unless exists $$val{"d"};
    $pval = $$val{"d"};
    my $res = jm_is_string($pval) && $pval eq "a";
    return 0 unless $res;
    return 0 unless exists $$val{"v"};
    $pval = $$val{"v"};
    return jm_is_array($pval);
}

sub _jm_obj_7($$$)
{
    my ($val, $path, $rep) = @_;
    return 0 unless jm_is_object($val);
    return 0 if jm_obj_size($val) != 2;
    my $pval;
    return 0 unless exists $$val{"d"};
    $pval = $$val{"d"};
    my $res = jm_is_string($pval) && $pval eq "o";
    return 0 unless $res;
    return 0 unless exists $$val{"v"};
    $pval = $$val{"v"};
    return jm_is_object($pval);
}


sub json_model_1($$$)
{
    my ($val, $path, $rep) = @_;
    my $res = jm_is_object($val);
    if ($res)
    {
        my $tag_0;
        if (defined($tag_0 = $$val{"d"}))
        {
            my $fun_0 = $_jm_map_0{$tag_0};
            $res = defined($fun_0) && &$fun_0($val, undef, undef);
        }
        else
        {
            $res = 0;
        }
    }
    return $res;
}


# initialization of global variables

our $initialized = 0;

sub check_model_init()
{
    if (!$initialized)
    {
        $initialized = 1;
        %_jm_map_0 = (
            "n" => \&_jm_obj_0,
            "b" => \&_jm_obj_1,
            "i" => \&_jm_obj_2,
            "u" => \&_jm_obj_3,
            "f" => \&_jm_obj_4,
            "s" => \&_jm_obj_5,
            "a" => \&_jm_obj_6,
            "o" => \&_jm_obj_7,
        );
        %check_model_map = (
            "" => \&json_model_1,
        );
    }
}

sub check_model_free()
{
    if ($initialized)
    {
        $initialized = 0;
        %_jm_map_0 = ();
        %check_model_map = ();
    }
}

sub check_model_mapper($)
{
    my ($name) = @_;
    die "unexpected model name \"$name\"" unless exists $check_model_map{$name};
    return $check_model_map{$name};
}

sub check_model($$$)
{
    my ($json, $name, $rep) = @_;
    die "unexpected model name \"$name\"" unless exists $check_model_map{$name};
    my $path = defined $rep ? [] : undef;
    return $check_model_map{$name}($json, $path, $rep);
}

#
# SCRIPT EXECUTION
#
check_model_init();
jm_main(\&check_model, \%check_model_map, JMC_VERSION);
check_model_free();

File: tests/mv-35/discs.java

import json_model.*;
import java.util.Map;
import java.util.Set;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

@SuppressWarnings("unchecked")
public class discs extends ModelChecker
{
    static public final String VERSION = "2";

    Map<Object, Checker> _jm_map_0_cmap;
    public Map<String, Checker> discs_map_pmap;

    public boolean _jm_obj_0(Object val, Path path, Report rep)
    {
        if (! json.isObject(val))
            return false;
        if (json.objectSize(val) != 2)
            return false;
        Object pval;
        if (! ((pval = json.objectValue(val, "d")) != null))
            return false;
        boolean res = json.isString(pval) && json.asString(pval).compareTo("n") == 0;
        if (! res)
            return false;
        if (! ((pval = json.objectValue(val, "v")) != null))
            return false;
        return json.isNull(pval);
    }

    public boolean _jm_obj_1(Object val, Path path, Report rep)
    {
        if (! json.isObject(val))
            return false;
        if (json.objectSize(val) != 2)
            return false;
        Object pval;
        if (! ((pval = json.objectValue(val, "d")) != null))
            return false;
        boolean res = json.isString(pval) && json.asString(pval).compareTo("b") == 0;
        if (! res)
            return false;
        if (! ((pval = json.objectValue(val, "v")) != null))
            return false;
        return json.isBoolean(pval);
    }

    public boolean _jm_obj_2(Object val, Path path, Report rep)
    {
        if (! json.isObject(val))
            return false;
        if (json.objectSize(val) != 2)
            return false;
        Object pval;
        if (! ((pval = json.objectValue(val, "d")) != null))
            return false;
        boolean res = json.isString(pval) && json.asString(pval).compareTo("i") == 0;
        if (! res)
            return false;
        if (! ((pval = json.objectValue(val, "v")) != null))
            return false;
        return json.isInteger(pval);
    }

    public boolean _jm_obj_3(Object val, Path path, Report rep)
    {
        if (! json.isObject(val))
            return false;
        if (json.objectSize(val) != 2)
            return false;
        Object pval;
        if (! ((pval = json.objectValue(val, "d")) != null))
            return false;
        boolean res = json.isString(pval) && json.asString(pval).compareTo("u") == 0;
        if (! res)
            return false;
        if (! ((pval = json.objectValue(val, "v")) != null))
            return false;
        return json.isInteger(pval) && json.asLong(pval) >= 0;
    }

    public boolean _jm_obj_4(Object val, Path path, Report rep)
    {
        if (! json.isObject(val))
            return false;
        if (json.objectSize(val) != 2)
            return false;
        Object pval;
        if (! ((pval = json.objectValue(val, "d")) != null))
            return false;
        boolean res = json.isString(pval) && json.asString(pval).compareTo("f") == 0;
        if (! res)
            return false;
        if (! ((pval = json.objectValue(val, "v")) != null))
            return false;
        return json.isDouble(pval);
    }

    public boolean _jm_obj_5(Object val, Path path, Report rep)
    {
        if (! json.isObject(val))
            return false;
        if (json.objectSize(val) != 2)
            return false;
        Object pval;
        if (! ((pval = json.objectValue(val, "d")) != null))
            return false;
        boolean res = json.isString(pval) && json.asString(pval).compareTo("s") == 0;
        if (! res)
            return false;
        if (! ((pval = json.objectValue(val, "v")) != null))
            return false;
        return json.isString(pval);
    }

    public boolean _jm_obj_6(Object val, Path path, Report rep)
    {
        if (! json.isObject(val))
            return false;
        if (json.objectSize(val) != 2)
            return false;
        Object pval;
        if (! ((pval = json.objectValue(val, "d")) != null))
            return false;
        boolean res = json.isString(pval) && json.asString(pval).compareTo("a") == 0;
        if (! res)
            return false;
        if (! ((pval = json.objectValue(val, "v")) != null))
            return false;
        return json.isArray(pval);
    }

    public boolean _jm_obj_7(Object val, Path path, Report rep)
    {
        if (! json.isObject(val))
            return false;
        if (json.objectSize(val) != 2)
            return false;
        Object pval;
        if (! ((pval = json.objectValue(val, "d")) != null))
            return false;
        boolean res = json.isString(pval) && json.asString(pval).compareTo("o") == 0;
        if (! res)
            return false;
        if (! ((pval = json.objectValue(val, "v")) != null))
            return false;
        return json.isObject(pval);
    }


    public boolean json_model_1(Object val, Path path, Report rep)
    {
        boolean res = json.isObject(val);
        if (res)
        {
            Object tag_0;
            if ((tag_0 = json.objectValue(val, "d")) != null)
            {
                Checker fun_0 = _jm_map_0_cmap.get(tag_0);
                res = fun_0 != null && fun_0.call(val, null, null);
            }
            else
                res = false;
        }
        return res;
    }


    public void init(JSON json)
    {
        if (!initialized)
        {
            try {
            _jm_map_0_cmap = new HashMap<Object, Checker>();
            _jm_map_0_cmap.put(json.safeJSON("\"n\""), new Checker() { public boolean call(Object o, Path p, Report r) { return _jm_obj_0(o, p, r);} });
            _jm_map_0_cmap.put(json.safeJSON("\"b\""), new Checker() { public boolean call(Object o, Path p, Report r) { return _jm_obj_1(o, p, r);} });
            _jm_map_0_cmap.put(json.safeJSON("\"i\""), new Checker() { public boolean call(Object o, Path p, Report r) { return _jm_obj_2(o, p, r);} });
            _jm_map_0_cmap.put(json.safeJSON("\"u\""), new Checker() { public boolean call(Object o, Path p, Report r) { return _jm_obj_3(o, p, r);} });
            _jm_map_0_cmap.put(json.safeJSON("\"f\""), new Checker() { public boolean call(Object o, Path p, Report r) { return _jm_obj_4(o, p, r);} });
            _jm_map_0_cmap.put(json.safeJSON("\"s\""), new Checker() { public boolean call(Object o, Path p, Report r) { return _jm_obj_5(o, p, r);} });
            _jm_map_0_cmap.put(json.safeJSON("\"a\""), new Checker() { public boolean call(Object o, Path p, Report r) { return _jm_obj_6(o, p, r);} });
            _jm_map_0_cmap.put(json.safeJSON("\"o\""), new Checker() { public boolean call(Object o, Path p, Report r) { return _jm_obj_7(o, p, r);} });
            discs_map_pmap = new HashMap<String, Checker>();
            discs_map_pmap.put("", new Checker() { public boolean call(Object o, Path p, Report r) { return json_model_1(o, p, r);} });
                super.init(json);
            }
            catch (Exception e) {
                throw new Error("cannot initialized model checker: " + e);
            }
        }
    }

    public void free()
    {
        if (initialized)
        {
            super.free();
            _jm_map_0_cmap = null;
            discs_map_pmap = null;
        }
    }

    public Checker get(String name)
    {
        return discs_map_pmap.get(name);
    }

    public Set<String> models()
    {
        return discs_map_pmap.keySet();
    }

    static public void main(String[] args) throws Exception
    {
        ModelChecker checker = new discs();
        Main.main("discs", checker, VERSION, args);
    }
}
